src/iceberg/avro/avro_reader.cc

```diff
@@ -91,16 +92,43 @@ class AvroBatchReader::Impl {
     // Create a base reader without setting reader schema to enable projection.
     auto base_reader =
         std::make_unique<::avro::DataFileReaderBase>(std::move(input_stream));
-    const ::avro::ValidSchema& file_schema = base_reader->dataSchema();
+    ::avro::ValidSchema file_schema = base_reader->dataSchema();
 
     // Validate field ids in the file schema.
     HasIdVisitor has_id_visitor;
     ICEBERG_RETURN_UNEXPECTED(has_id_visitor.Visit(file_schema));
+
     if (has_id_visitor.HasNoIds()) {
-      // TODO(gangwu): support applying field-ids based on name mapping
-      return NotImplemented("Avro file schema has no field IDs");
-    }
-    if (!has_id_visitor.AllHaveIds()) {
+      // Apply field IDs based on name mapping if available
+      if (options.name_mapping) {
+        MappedField mapped_field;
+        // Convert NameMapping to MappedFields for nested mapping
+        mapped_field.nested_mapping =
+            std::make_shared<MappedFields>(options.name_mapping->AsMappedFields());
+        ICEBERG_ASSIGN_OR_RAISE(
+            auto new_root_node,
+            CreateAvroNodeWithFieldIds(file_schema.root(), mapped_field));
+
+        // Create a new schema with the updated root node
+        auto new_schema = ::avro::ValidSchema(new_root_node);
+
+        // Verify that all fields now have IDs after applying the name mapping
+        HasIdVisitor verify_visitor;
+        ICEBERG_RETURN_UNEXPECTED(verify_visitor.Visit(new_schema));
+        if (!verify_visitor.AllHaveIds()) {
```

Comment:
I think we can remove this extra check now. Your implementation can guarantee all fields are processed.